crab/agents/backend_models/camel_model.py

```diff
@@ -0,0 +1,158 @@
+    def __init__(
+        self,
+        model: str,
+        model_platform: str,
+        parameters: Optional[Dict[str, Any]] = None,
+        history_messages_len: int = 0,
+    ) -> None:
+        if not CAMEL_ENABLED:
+            raise ImportError("Please install camel-ai to use CamelModel")
+        self.parameters = parameters or {}
+        # TODO: a better way?
+        self.model_type = find_model_type(model)
+        self.model_platform_type = find_model_platform_type(model_platform)
+        self.client: Optional[ExternalToolAgent] = None
+        self.token_usage = 0
+
+        super().__init__(
+            model,
+            parameters,
+            history_messages_len,
+        )
+
+    def get_token_usage(self):
+        return self.token_usage
+
+    def reset(self, system_message: str, action_space: Optional[List[Action]]) -> None:
+        action_schema = self._convert_action_to_schema(action_space)
+        config = ChatGPTConfig(
+            tool_choice="required", tools=action_schema, **self.parameters
+        )
+        backend_model = ModelFactory.create(
+            self.model_platform_type,
+            self.model_type,
+            model_config_dict=config.as_dict(),
+        )
+        sysmsg = BaseMessage.make_assistant_message(
+            role_name="Assistant",
+            content=system_message,
+        )
+        self.client = ExternalToolAgent(
+            model=backend_model,
+            system_message=sysmsg,
+            external_tools=action_schema,
+            message_window_size=self.history_messages_len,
+        )
+        self.token_usage = 0
+
+    # TODO: convert Action into OpenAIFunction
+    @staticmethod
+    def _convert_action_to_schema(
+        action_space: Optional[List[Action]],
+    ) -> Optional[List[OpenAIFunction]]:
+        if action_space is None:
+            return None
+        return [
+            OpenAIFunction(
+                action.entry,
+            )
+            for action in action_space
+        ]
+
+    @staticmethod
+    def _convert_tool_calls_to_action_list(tool_calls) -> List[ActionOutput]:
+        if tool_calls is None:
+            return tool_calls
+
+        return [
+            ActionOutput(
+                name=call.function.name,
+                arguments=json.loads(call.function.arguments),
+            )
+            for call in tool_calls
+        ]
+
+    def chat(self, messages: List[Tuple[str, MessageType]]):
+        # TODO: handle multiple text messages after message refactoring
+        image_list: List[Image.Image] = []
+        content = ""
+        for message in messages:
+            if message[1] == MessageType.IMAGE_JPG_BASE64:
+                image = decode_image(message[0])
+                image_list.append(image)
+            else:
+                content = message[0]
+        usermsg = BaseMessage.make_user_message(
+            role_name="User",
+            content=content,
+            image_list=image_list,
```

Comment:
Is it possible to make the images between two text content like OpenAI API does?

Reply:
Not available for now, but CAMEL is refactoring the message and this will be available afterwards

Reply:
Sure. Let's wait